File: TypeScript/ExampleTests/common/AdderTyped.ts
import { AdderString } from "./AdderString.js";
import * as _json from "./json.js";

export class AdderTyped {
  number1: number;
  number2: number;
  result: number;

  constructor(number1: number, number2: number, result: number) {
    this.number1 = number1;
    this.number2 = number2;
    this.result = result;
  }

  static fromStringObj(s: AdderString): AdderTyped {
    return new AdderTyped(
      s.number1 !== "" ? Number(s.number1) : 0,
      s.number2 !== "" ? Number(s.number2) : 0,
      s.result !== "" ? Number(s.result) : 0
    );
  }

  toStringObj(): AdderString {
    return new AdderString(
      String(this.number1),
      String(this.number2),
      String(this.result)
    );
  }

  static toStringList(list: AdderTyped[]): AdderString[] {
    return list.map(t => t.toStringObj());
  }

  static fromStringList(list: AdderString[]): AdderTyped[] {
    return list.map(s => AdderTyped.fromStringObj(s));
  }

  toJsonValue(): Record<string, unknown> {
    return {
      number1: this.number1,
      number2: this.number2,
      result: this.result,
    };
  }

  toJSON(): string { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m: unknown): AdderTyped {
    return new AdderTyped(
      _json.asInt(_json.requireField(m, "number1"), "number1"),
      _json.asInt(_json.requireField(m, "number2"), "number2"),
      _json.asInt(_json.requireField(m, "result"), "result")
    );
  }

  static fromJSON(text: string): AdderTyped {
    return AdderTyped.fromJsonValue(_json.parse(text));
  }

  static toJSONList(items: readonly AdderTyped[]): string {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text: string): AdderTyped[] {
    const raw = _json.asArray(_json.parse(text), "AdderTyped") ?? [];
    return raw.map((e) => AdderTyped.fromJsonValue(e));
  }

  toString(): string {
    return `number1=${this.number1}, number2=${this.number2}, result=${this.result}`;
  }

  equals(other: AdderTyped): boolean {
    return this.number1 === other.number1
      && this.number2 === other.number2
      && this.result === other.result;
  }
}

File: Python/ExampleTests/common/adder_typed.py
from . import json_util as _json
from .adder_string import AdderString

class AdderTyped:
    def __init__(self, number1: int = 0, number2: int = 0, result: int = 0):
        self.number1 = number1
        self.number2 = number2
        self.result = result

    @classmethod
    def from_string_obj(cls, s: AdderString) -> 'AdderTyped':
        return cls(
            int(s.number1) if s.number1 else 0,
            int(s.number2) if s.number2 else 0,
            int(s.result) if s.result else 0
        )

    def to_json_value(self) -> dict:
        return {
            'number1': self.number1,
            'number2': self.number2,
            'result': self.result,
        }

    def to_json(self) -> str:
        return _json.dumps(self.to_json_value())

    @classmethod
    def from_json_value(cls, m: dict) -> 'AdderTyped':
        return cls(
            _json.as_int(_json.require(m, 'number1'), 'number1'),
            _json.as_int(_json.require(m, 'number2'), 'number2'),
            _json.as_int(_json.require(m, 'result'), 'result')
        )

    @classmethod
    def from_json(cls, text: str) -> 'AdderTyped':
        return cls.from_json_value(_json.loads(text))

    @staticmethod
    def to_json_list(items) -> str:
        return _json.dumps([item.to_json_value() for item in items])

    @classmethod
    def from_json_list(cls, text: str) -> list:
        raw = _json.as_list(_json.loads(text), 'AdderTyped')
        return [cls.from_json_value(e) for e in raw]
